states/states.go

```diff
@@ -263,6 +263,8 @@ type SaveImage struct {
 	// list (usually used for multi-platform setups). This means that the image
 	// can only be a single-platform image.
 	NoManifestList bool
+
+	Platform platutil.Platform
```

Comment:
This feels incorrect - IIRC we should never have bare platforms here because we can't distinguish between null platform that resolves to the native platform vs null platform that resolves to something else. We should always use a `PlatformResolver` instead of a bare platform in these cases. IIRC we already have a `PlatformResolver` in the `states.SingleTarget` struct that contains the `SaveImage`...

Though now `SaveImage` could be loose, only part of a wait block...

I'm now a bit worried about what happens if we attempt to save an image without having run `FinalizeStates` for the target that created it. Note how the `PlatformResolver` gets assigned at this point: https://github.com/earthly/earthly/blob/main/earthfile2llb/converter.go#L1375.

I feel that without `FinalizeStates` there will be many edge cases where things will not work as expected.

Could we instead not wait on `SAVE IMAGE` itself, but rather on things like `BUILD` statements (and of course, waiting on `BUILD` will trigger the save image exports if any)? Then we could ensure that `FinalizeStates` has run for the target being invoked (and then we could use the `PlatformResolver` from the `mts.Final`).

Reply:
> Could we instead not wait on SAVE IMAGE itself, but rather on things like BUILD statements

This might not work for cases where there is no `BUILD`, e.g.

```
VERSION 0.6
FROM alpine
RUN echo generate-data > output
SAVE IMAGE --push alexcb132/testing:tag
RUN --push echo do this after
```

Which should be translated to:
```
VERSION --wait-block 0.6
FROM alpine
WAIT
    RUN echo generate-data > output
    SAVE IMAGE --push alexcb132/testing:tag
END
RUN --push echo do this after
```

Given it's a pattern we already support, it would be nice if we could continue to support it.

As we discussed on Friday, I'll first try including a bool to detect if it's a single image for the current worker's platform (for cases no `--platform` was given), or a pseudo-multiplatform image which only has a single platform defined (for cases `--platform` was passed in).